### DISCUSSION TITLE
Service calls with return values

### DISCUSSION THREAD

**Discussion by allenporter** [home-assistant/architecture#777] [Entity Models] [answered]
*Back in April or so there was a discord discussion about this, so thought it was worth following through on*

## Proposal

Support return values for services.

## Use cases

Example use cases:
- Send an email with a list of birthdays today from my calendar
- Picnic search results https://github.com/home-assistant/core/pull/67877
    - Edit: Return all the items of the shopping list, for use in another service eg a notification.
- An expensive Cloud API call used in an automation
    - Edit: Compute distance matrix / travel time on demand, rather than polling in the background

## Best practices and limitations

- Failures should throw exceptions like they do today, and not return failure status codes.  Never use with batching and try to support partial failure as it is easy to get wrong.
- Return value is a dict. The returned fields must be described in `services.yaml`
- Do not use return values from services when other simpler approaches will work e.g. would not replace returning state for a sensor, or return state for a bunch of sensors in the integration, etc.
- For entity services: No return values when matching multiple entities
- Non-blocking service calls obviously won't get return values

## Additional detail

Data from a services can be returned from an action during an automation, and set as variables that can be used in later automations.  e.g. the data isn't necessarily associated with an existing trigger, as the trigger may be something else entirely.

**Comment by MartinHjelmare** [comment]
Use case:
Return all the items of the shopping list, for use in another service eg a notification.

**Reply by ScottG489** [reply]
Wouldn't the idea here just be to have an entity which contains all items of the shopping list and you can just get that entity's value for this data?

**Comment by frenck** [comment]
I've been thinking about these things for a long time as well. GitHub Actions & GitLab CI are nice examples of handling return values in some way from their "service calls"/"actions".

The problem is: If you want to consume those return values, you will always need templates.

Which makes me think: If You need templates anyways, why not aim for allowing template extensions/methods by integrations?

Or, in case we do allow return values on services as variables as you propose, what does that mean fo existing template methods? For example, looking up a device or area ID, would that become a service?

**Reply by allenporter** [reply]
Leaning into templates:  Great point. I agree, there are no cases other than consumption via template that I'm aware of. Staying within templates could be more straight forward. Not sure if async from templates is risky, but looks like its supported. Is that a non-starter or reasonable to explore?

I'm not sure why, but the existing methods provided by templates seem too simple/straight forward to be services. Maybe it's just because they are simple implementations with simple arguments. They also have different return values, not just a dict.  

With respect to convergence for services, it also seems related to websocket and rest APIs? I am not sure what the principles are with respect to when something is a rest api vs a websocket (though imagine there are security considerations for rest that are special) but often many have similar functionality (e.g. the various versions of camera  websockets or rest apis). My assumption is that the service description language is used to power the service call UI, but it seems interesting to codify/describe rest APIs too. Maybe converge on services there, then have options to enable services vs rest vs websocket.

**Reply by allenporter** [reply]
I've only limited experience with templates, but I think the main considerations are:
- Integration must be isolated from each other and other template globals
- Only specific functions are allow-listed.
- Return values should also be limited in scope (e.g. simple values, dataclass with a finite set of attributes, etc)

So the simplest possible API I can think of is for some `DOMAIN` have an platform `template.py`:
```
# Allow listed set of functions exported as `DOMAIN.<function name>(...)`
TEMPLATE_FUNCTIONS = {
    "base64_encode",
    "base64_encode_service",
}


def base64_encode(hass: HomeAssistant, value: str) -> str:
    """Perform base64 encode."""
    return base64.b64encode(value.encode("utf-8")).decode("utf-8")

async def base64_encode_service(hass: HomeAssistant, value: str) -> str:
    """Perform base64 encode using async service."""
    encode_service = hass.data[DOMAIN]["encode_service"]
    return await encode_service.send(value)
```

So then from a template we can call these like:
-`{{ my_domain.base64_encode("example") }}`
- `{{ my_domain.base64_encode_service("example") }}`

My intuition is that it would only need to support function calls, and not exposing other types of objects, attributes, or constants. New template functions would not be a replacement for other existing ways to use entities, but could perform extended operations with an entity as input like the calendar event example.

I have only partially looked at the implementation, but the biggest issue where would be dealing with async. Today it appears the templates are created and rendered within a sync context, so it might be a bit of a rewrite to use the async jinja2 APIs.  (Templates are currently rendered from a synchronous method inside the event loop.)

**Reply by balloob** [reply]
Templates are being rendered as a callback function. This means that it is not possible currently to do async stuff in it. I'm also not sure if we would ever want to. It means we would either a) rewrite all places that render templates (impossible imo) or b) update some places to allow async functions (a bit like we have a limited template environment for triggers).

I think that I would leave the distinction between services and templates to be what is provided by integrations vs provided by HA core/helpers.

I also don't think that we always need templates for this. Yes, for it's raw value that's indeed the case. But for example, in the action UI we already extract "play media" as a separate action. Maybe we could come up with some services with return values that we would like to promote, like fetching/searching calendar events, get shopping list items or query an LLM.

**Reply by allenporter** [reply]
I am leaning against templates as async + templates does seem to be a little wild. I did see `enable_async` in jinja2, but it looked like quite a big lift to rewrite the templates to actually use it. 

Since we're talking about this again, for completeness here are some prototypes i've implemented in the last ~year:
-  [template platform prototype](https://github.com/home-assistant/core/compare/dev...allenporter:template-platform?expand=1) which only works synchronously only. I misspoke before thinking i had async working. Even for the sync example, i think this is now broken after some refactors since initialization order changed.
-  [Service return values prototype](https://github.com/home-assistant/core/compare/dev...allenporter:home-assistant-core:service-return-values?expand=1) is straight forward, adding return values where they are missing.
- [Data sources prototype](https://github.com/home-assistant/core/compare/dev...allenporter:home-assistant-core:datasource?expand=1) for a standalone query abstraction based on the [Data Source proposal doc](https://docs.google.com/document/d/1-UGjbGkwCO12e-krcO-TfnAOuvjMhwGRJYQlexcGSCw/edit#heading=h.bws5tog1aiwc). This works and I believe I was able to use it in an automation.
- As a workaround, you can use a  [Rest sensor that queries the calendar Rest API](https://community.home-assistant.io/t/calendar-new-entitites-or-helper-that-show-a-list-of-calendar-events-within-a-certain-scope/358014/28) which works for Calendar since it happens to have a rest API.  This is fairly silly since you set up an API token, then use a rest sensor to call the home assistant calendar rest API to get the data. Could be streamlined, but i'm not sure that we're excited about adding more rest APIs. This makes the data always available in an entity, rather than called during a script, which is another difference.

**Reply by balloob** [reply]
My current leaning is towards adding return values to services prototype.

One hard requirement should be that the return values are all JSON serializable (including HA extensions for datetime). We wouldn't want to open it up to return objects that can call functions. It needs to be limited to data (objects with as_dict functions would be ok).

From a quick scan, we should also add the following features:
 - call service action step (used in scripts/automations) should store result in a variable so it can be accessed in subsequent steps. This is a breaking change since the result variable could collide with what a user is using today
 - `services.yaml` + validation should support describing this.

**Reply by OnFreund** [reply]
I think it can be really helpful if failures can also be communicated as events. Return values are great, but sometimes I want to centrally handle all failures pertaining to a device, whether it was through an automation, script, or the UI.
For example, I'd like to be notified every time there's a failure to arm the alarm, even if the attempt is through the UI.

**Reply by balloob** [reply]
that's out of scope for this architecture discussion.

**Reply by OnFreund** [reply]
This was already raised as an alternative [earlier in this discussion](https://github.com/home-assistant/architecture/discussions/777#discussioncomment-3109263).

**Reply by PeteRager** [reply]
Events get fired when a service call occurs, so it would make sense to also fire an event when a service call completes.  This would allow for calculating statistics on service call latency, failure rates, automations taking actions, etc.  That said I agree with @balloob that we should leave it out of the scope of this discussion - as this new event could be implemented today regardless of the outcome of adding return values to service calls.

**Reply by ScottG489** [reply]
Isn't this functionality already available via state changes/events? Meaning, instead of getting a return value from a service, the service call would result in an update to the state of an entity and that would be your "return value", or just wait for an event to be emitted with your return value.

So rather than the usage flow suggested here to call a service and wait for the return value, you'd call the service and wait for an entity's state to update or wait for an event to be fired which contains your "return value".

However, implementing something like this now in an automation or w/e would be pretty heavy handed. I've definitely felt myself wanting functionality like this in the past. So it's not to say we can't make changes to make this more seamless for users, but keeping it event based seems like it would align more with core's desired/existing architecture. Thoughts?

**Reply by allenporter** [reply]
> Isn't this functionality already available via state changes/events? Meaning, instead of getting a return value from a service, the service call would result in an update to the state of an entity and that would be your "return value", or just wait for an event to be emitted with your return value.


Good question. The part missing from the proposal is the motivation that these things above are not usually allowed to be written to the state machine. They are large (e.g. list of events over some time range) or dynamic and potentially changing, or richer data than we'd allow in the state attributes (repeated lists or dictionaries)

**Reply by OnFreund** [reply]
Good point, but that state machine doesn't capture the error. Additionally, service calls don't necessarily map to state changes (e.g. calling a notify service)

**Reply by PeteRager** [reply]
It needs to be deterministic.  Meaning if the same service is called from different automations simultaneously (or close to), the return values should be independent and not overwritten - which would be the case if the result is stored in the state machine.

**Reply by ScottG489** [reply]
> The part missing from the proposal is the motivation that these things above are not usually allowed to be written to the state machine

That's understandable that we wouldn't want it in the state machine. However, the idea to use an emitted event to contain the "return" data could still work, no?

> Good point, but that state machine doesn't capture the error.

As for errors, balloob mentioned above that might be out of scope. I think it may be an important consideration, but at least for a first iteration I don't see why we couldn't continue to handle errors the same way we do currently - the error is logged.

Thinking about it more now, using an entity's state as a holder for the return value does seem like it's overloading the purpose of an entity a bit much. Similarly to what we were doing with persistent notifications which was removed recently.

I still think an emitted event could work though. The flow I'm imagining is you'd make the service call and then immediately (or before) make a trigger that listens for the event representing the result from this service call.

As I type this I'm actually noticing below that it looks like @depoll may have been [suggesting a similar architecture](https://github.com/home-assistant/architecture/discussions/777#discussioncomment-3109263)?
Edit: It looks like this may also be the pattern that was originally implemented in the PR mentioned in the first comment of this discussion: https://github.com/home-assistant/core/pull/67877#discussion_r888284392

**Reply by PeteRager** [reply]
The event could work as long as it could be correlated with a specific instance of a service call.  So for example if a service call is given a guid and that guid is in the event, then the receiver could distinguish if the event was for its service call vs someone else invoking the same service.

Said another way, the return value must be from a specific instance of a service call.

**Reply by ScottG489** [reply]
Yes I was thinking that as well. That's the one thing that could require some change to the service call design I think. I'm curious to hear some core devs opinions on this implementation idea, though.

**Reply by allenporter** [reply]
The service call + event can definitely work from a technical perspective, but it seems difficult to use IMO. It has the advantage of not introducing a new concept, but seems difficult to describe and document. With a service with return values (or data sources) it is easier to hook into the UI (e.g. imagine the UI could run the serivce call and show example return values in the UI editor).

Outside of home assistant i think it is pretty rare to see the design pattern that RPCs that publish events for consumption, though I have seen it for things like long lived processing e.g. processing a video, or some other kind of work queue or something

**Reply by ScottG489** [reply]
Those are some good points.

I think I was imagining the pattern as the architecture and not necessarily the method of use. It would definitely be cumbersome for the user to have to call the service and then listen for events themselves.

Looking at your prototype though, I'm starting feel Paulus' assessment was right. I think we should assign it to a variable which is a pattern we have elsewhere like for triggers.

I wonder if, in addition to this, there would still be value in emitting an event with the result. Perhaps for longer running service calls where it would be better for them to be async (e.g. `hassio.backup_*` services)?

**Reply by OnFreund** [reply]
In my opinion the main motivations for an event based approach are:
1.  Centralized error handling
2. Regardless of context

For example, have an automation that handles a failure to arm the alarm, whether that arming attempt was through the UI, script, automation, etc...
Without events, the only way to handle errors that originate from the UI would be to create a template alarm control panel, with services that encapsulate error handling. That's very cumbersome for one entity, let alone having to do that for each entity.

**Reply by allenporter** [reply]
That seems like a fairly different goal,.outside of the scope of this proposal (stated a couple times error handling is explicitly not a goal of what we're discussing here). I'd encourage a separate discussion.

**Comment by depoll** [comment]
@allenporter linked me here (thanks!) from discord after I optimistically opened a PR with a slightly different take on how to get there: https://github.com/home-assistant/core/pull/74729

I started from using wait/repeat/etc. as prior art. While the mechanism for a service reporting its result in my PR is firing an event, the way it gets used in a script/automation is by setting a variable as a byproduct of the service call. Just as you have `{{wait.trigger}}` you can also have `{{service.result}}` after calling a service.

There are a few nice things about this: it doesn't require any particular extensions to templates that may or may not work depending on the context, it matches the kind of side effects other actions have on variables in scripts, and it allows any service (including scripts that a user may be building for themselves) to without having to drop into code.

It doesn't currently solve the "return type of part of the schema of the service" bit, but that seems at least possible to address.

Curious if anyone has thoughts on this approach? I'm new to contributing here, so it's entirely possible there's some better place for this that I'm totally ignorant to 😉

**Reply by DanielBaulig** [reply]
Just chiming in here, because I recently thought about this myself a little and realized that you can actually already do this with some really out there automation code by simply utilizing events, parallel and waits. 

For a "global function" you have an automation that runs at startup and then just busy-spins while waiting for an event to fire on every iteration. It then, in response to the event, just executes some actions and returns any resulting value by firing a response event.

Any automations wanting to utilize the "global function" can then simply fire the event and wait for the resulting event. To allow for concurrent calls an identifier can be provided upon call, that can be matched against on the return path. 

To have a "local function" you can do the same but spin up your busy loop from within a `parallel` action in the automation you want to make the "local function" available in. You'd probably have to explicitly `stop:` at the end of the automation to terminate the parallel busy spinning branch.

Obviously this is super hacky and I wouldn't recommend doing this to anyone, but technically, users can already do async service calls with return values in the style that @depoll suggests. They just have to go there on knees and hands. And there isn't really anything stopping integrations from doing the same if they so choose to.

**Reply by OnFreund** [reply]
I like the event-based approach, because it doesn't require the service to be called from a script/automation in order to obtain its return value. I was about to open a new discussion about being able to trigger based on service call failures (e.g. notify me when there's a failed attempt to arm the alarm), but I think an event-based approach, at the generic service level, rather than individual service, addresses this use case.

**Reply by jlsjonas** [reply]
This is honestly where the node-RED integration currently shines (as a viable intermediary workaround), would be great if we could cover more of those cases natively, especially once integrations start taking advantage of it!

**Comment by PeteRager** [comment]
Currently it seems like service calls that control device are “suggestions” and outside of returning that the suggestion was accepted there is no real feedback on whether the service call did what was expected.

Two specific examples:

Zwavejs - setting thermostat hvac_mode.  Automation calls the service, zwavejs integration then calls into the zwavejs websocket with the command and then returns success which is retuned to the automation. After that,  if the command times out in zwavejs an error is logged in its log file and that is it.

Nest - setting the thermostat temperature.  Automation calls the service, nest integration sends it to googles api and returns the apis result.  Later on google delivers this to NEST.   Sometimes it never gets delivered.  I monitor this it’s ~0.1% failure rate.

And yes, if googles code was perfect and zwave networks and devices worked perfectly neither of these would be issues.  While I love striving for perfection, sometime the best way for perfection is to make the system resilient to imperfection. 

As a result, I’m now implementing “reliable service calls”, that issue the command, check for the change, retry if necessary and log messages on retry failure.  This works well on the two above examples since they seems to be transitory and usually success on the second try.  Doing this required a stack of HASS scripts, shell scripts to blast out the config, etc.

Here’s the thing.  The retry logic I describe is what most average non technical users would expect to happen.  If I have an automation to turn lights off at 10pm, then it should just work.  If I tell my nest to warmup the house at 6, it should just work.

**Comment by ve6rah** [comment]
@PeteRager I think this is one of the biggest weaknesses of ha right now. I have most of my automations include checks to see if the action I called actually happened, but it makes for clunky automations and certainly isn't what a new user would expect they need to do. It changes all automations from "if this, then that" to "if this, then that, then check that, if not done try again, check that, if still not done notify user", and that doesn't even get into all the automations I have solely to check to make sure every device is still online!

But you only want to come home to find your door didn't lock on command so many times! (I find the failure rate on that one to be at least an order of magnitude higher than the 0.1% you quoted. I'd guess 2-5% with my zwave lock)

**Reply by balloob** [reply]
This is not going to be solved by return values. It won't be able to include any changed states because devices that push state updates might end up pushing after the service call is done. If a service fails, it raises an error so you can already do error checking if the server was a success, you don't need a return value for that.

**Reply by PeteRager** [reply]
That is not universally true.   For example in zwavejs, if the message gets destroyed due to RF, while zwavejs generates a timeout error and logs it; this is AFTER the integration has already returned success to the automation.  Do you consider this a bug?

When an integration does return an error this will cause the automation to fail unless continue-on-error is true.  However there is no way in the automations next step to determine that the prior step failed.

My consideration in posting this in this topic, was that if we want to have return values from service calls as part of that design change perhaps there is a way to address the retry issue and/or provide a simpler mechanism.   I have it solved for me with a 100 line script.  I think there’s a lot of people solving it with similar scripts and alot of non-sophisticated users posting in the forums wondering why there lights don’t always turn off.

**Reply by balloob** [reply]
If Z-Wave JS can detect that sending a message has failed, than it should report that back to the caller, yes. Return values for services wouldn't help here either, as you're still having a SUCCESS result? 

Errors should raise and bubble up. It's up to the place that calls a service (script, automation Python etc) to decide if an error can/should be ignored.

**Reply by OnFreund** [reply]
Is there a way to know that an action errored? Because I couldn't find anything [in the docs](https://www.home-assistant.io/docs/scripts/#continuing-on-error).

**Reply by PeteRager** [reply]
No, since there is no return value from the service call in the automation (in this case an exception) there is no current way to tell.  If each automation step had return context (step_name.error, step_name.return_value, etc.) that would be helpful.  If the error return codes were standardized (like the http ones), then you could tell if it’s a timeout error, auth error, server busy. Then it’d be easy to create a shim layer between the automation and the integration to automatically retry on specific error code.

**Reply by allenporter** [reply]
Error handling is a complex topic and difficult to get right, and I think it would be more harmful than helpful to try to expose this level of detail. E.g. I don't think you can handle those error scenarios differently an automation. For example, a timeout happens, but the call still may have succeeded, and this detail didn't really help and you have to re evaluate the state of the system anyway.

This proposal explicitly stated that it is intentionally not changing how errors are handled and avoiding topics like partial failure.  That is a very different topic than service return values, and not something I think should pull into this proposal.

**Reply by PeteRager** [reply]
Valid points.  Here are my thoughts on what could be included from a user perspective. 

a) Each automation step may have a return value that is accessible via Jinja from any subsequent step of the automation. To get this scoped properly may require adding a step-ID or sluggfying the alias. 

```
{{ mystep1.retval.atrib }}
{{mystep2.retval.attribute }}
```

b) Support simple return values (strings, number) and dictionaries.   This way the return value of a service that returns True or False can be directly used in Jinja w/o added complexity. 

```
{% if mytep1.retval == True %}
```

c) For consistency consider adding a value to the step object that captures the returned exception 😃

```
{% if mytep1.error is not None %}
```

d) scripts should be able to return values.  Scripts are very useful and powerful.  They should be able to provide a return value (just like a service call) into the calling step.

**Reply by PeteRager** [reply]
> @PeteRager I think this is one of the biggest weaknesses of ha right now. I have most of my automations include checks to see if the action I called actually happened, but it makes for clunky automations and certainly isn't what a new user would expect they need to do. It changes all automations from "if this, then that" to "if this, then that, then check that, if not done try again, check that, if still not done notify user", and that doesn't even get into all the automations I have solely to check to make sure every device is still online!
> 
> But you only want to come home to find your door didn't lock on command so many times! (I find the failure rate on that one to be at least an order of magnitude higher than the 0.1% you quoted. I'd guess 2-5% with my zwave lock)

I’ve started a project to address this have an initial version here.   The strategy will be to have a “reliable services” integration that sits as a layer above existing services and does the retry coordination, poll after set that some devices need, etc.  First draft is it generates a set of HASS script wrappers, second generation will be a python integration.  

https://github.com/PeteRager/hass_rs

**Comment by balloob** [comment, marked as answer]
I had a call with @frenck, @MartinHjelmare and @emontnemery today and we arrived at the following:

**Opt-in return values**

Users need to opt-in to return values for every service call. This allows us to keep the system fast, as for certain calls we will not care about return values.

- The `ServiceCall` class will get a new property to indicate a return value was requested.
- Callers need to pass to `hass.services.async_call` if they want a return value. It defaults to False.

**Return values need to be a dictionary that is fully JSON serializable**

All return values need to be dictionary containing the results. This allows future extension without breaking changes.

**Return values need to be documented in services.yaml**

Services.yaml will need to describe the shape of the return values. This should also be added to the schema of the script integration. This allows us to guide the user in understanding the data.

**Scripts get return values by extending the `stop` action.**

The [Stop action](https://www.home-assistant.io/docs/scripts/#stopping-a-script-sequence) will be extended with a new `result` field to define the result for the whole script. The field value is interpreted as a Jinja2 template.

**Calling a service in Script syntax will require opt-in to get the result.**

The [Call Service action](https://www.home-assistant.io/docs/scripts/#call-a-service) will get a new key `result_variable` which has a string as value. If set, the service is called requesting a return value, and it will be stored in the variable as defined in `result_variable`.

**Errors are thrown, not returned in the result.**

Any errors need to be continued to raise as exceptions.

**Comment by RogerSelwyn** [comment]
Sorry to intrude on an architecture discussion. 

I've just noticed that the change implemented only supports entity_components and does not include entity_platforms. I've tried making what I thought was a simple change to core, but can't get that to work. Was it specifically intended to exclude platforms?

**Reply by frenck** [reply]
For now, this has been excluded, as we need to handle multiple return values from multiple targets in that case. One of the considerations is to only make it work if there is a single target.

**Reply by RogerSelwyn** [reply]
Thanks for letting me know

**Comment by oskars-viksna** [comment]
Hello all!
Just crashed into critical blocker for my HA use case - this new way with returning value from service calls are not supported at all when service is called via API. Checked source code and it clearly is there - http wrapper has no way to pass required return_response variable. Unfortunately some integrations are switching to this new way (reported issue against OpenWeatherMap about this https://github.com/home-assistant/core/issues/121219) and because of that are becoming useless for using via API. Please fix this ASAP in http wrapper or notify all integration owners to avoid using this way for data retrieval.
Thanks.

**Reply by frenck** [reply]
This is not the place to report issues or place feature requests.

This part has not been in scope for this architectural proposal.

../Frenck
